frontend/src/tree/TreeItemHandlersContext.tsx

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Obeo.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import React from 'react';
+import { TreeItemType } from './TreeItem.types';
+import { TreeItemHandler } from './TreeItemContextMenu.types';
+
+// Catch-all, used when no custom handler is found for an item.
+const defaultItemHandler: TreeItemHandler = {
+  handles: (item) => true,
+  getModal: (name) => null,
+  getMenuEntries: (item) => {
+    return [];
+  },
```

Comment:
Note for myself, I think an API like that could do the trick. In the code of the application using Sirius Components:

```
const canHandle = (item: GQLItem) => {
  return item.kind.startsWith('siriusComponents://representations?type=diagram');
};

const DiagramContextMenuContribution = ({ onClose } : ContextMenuContributionProps) => {
  const [dialogVisible, setDialogVisible] = useState(false);

  const dialog = dialogVisible ? (
    <Dialog open={true} onClose={onClose}>
      <DialogTitle id="dialog-title">Clone the diagram</DialogTitle>
      <DialogContent>
        <DialogContentText>
          Clone the diagram, as expected
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Clone</Button>
      </DialogActions>
    </Dialog>
  ) : null;

  return (
    <>
      <MenuItem onClick={() => setDialogVisble(true)}>
        <ListItemIcon>
          <AddIcon />
        </ListItemIcon>
        <ListItemText primary="Clone diagram with dialog" />
      </MenuItem>
      <MenuItem onClick={() => { /* do something and then */ onClose() }}>
        <ListItemIcon>
          <AddIcon />
        </ListItemIcon>
        <ListItemText primary="Clone diagram without dialog" />
      </MenuItem>
      {dialog}
    </>
  );
};

<WorkbenchContribution>
  <ContextMenuContribution canHandle={canHandle} component={DiagramContextMenuContribution}/>
  <Workbench />
</WorkbenchContribution>
```

With an API like that one, a contribution only needs to tell us when it should apply just like your `handles` function but this time with a simple props and it should give us a component. This component can do anything it want with regard to its lifecycle, such as displaying anything, opening a dialog or even multiple dialogs and wizards, etc. It can communicate with a server etc. Sirius Components does not need to care about that (especially the dialog parts). We just give it the way to tell us that after an interaction with the end user it has done its job thanks to the props `onClose`. We may need to give it additional props but that's the core behavior. With such approach, the API would be way less Java-like (see how your TreeItemHandler are structured) and more React-like. It would also let contributions uses more easily React lifecycle hooks etc.

It feels closer to the API of React Router for example. I still need to figure out some details but I think it would be more powerful for our users since it would offer more flexibility.

Reply:
Or even better, something like:

```
<Workbench>
  <ContextMenuContribution canHandle={canHandle} component={DiagramContextMenuContribution} />
  <RepresentationContribution ... />
</Workbench>
```

This way, the React context may not even be used anymore... I'll have to try something like that, it would open the door to a standard way to contribute to anything in the workbench.